syntax = "proto3";

service MemoryManager {
  rpc Create (CreateRequest) returns (CreateResponse);
  rpc Set (SetRequest) returns (Empty);
  rpc Get (GetRequest) returns (GetResponse);
  rpc IncreaseRefCount (IdRequest) returns (Empty);
  rpc DecreaseRefCount (IdRequest) returns (Empty);
}

message CreateRequest {
  uint32 size = 1;
  string type = 2;
}

message CreateResponse {
  uint32 id = 1;
}

message SetRequest {
  uint32 id = 1;
  string value = 2;
}

message GetRequest {
  uint32 id = 1;
}

message GetResponse {
  string value = 1;
}

message IdRequest {
  uint32 id = 1;
}

message Empty {}
